### ISSUE TITLE
Large document updates resulting in stalled tasks with MDB_TXN_FULL: Transaction has too many dirty pages

### ISSUE THREAD

**Issue by dtomlinson91** [meilisearch/meilisearch#3603]
**Describe the bug**
Meilisearch is failing to update documents after updating to Meilisearch v1.0.2. The `documentAdditionOrUpdate` tasks are seemingly stuck in processing then eventually fail after a long period (6 days) with the following error:

```json
{
    "canceledBy": null,
    "details": {
        "indexedDocuments": 0,
        "receivedDocuments": 10000
    },
    "duration": "P6DT60514.509414643S",
    "enqueuedAt": "2023-03-14T05:22:47.47082584Z",
    "error": {
        "code": "internal",
        "link": "https://docs.meilisearch.com/errors#internal",
        "message": "internal: MDB_TXN_FULL: Transaction has too many dirty pages - transaction too big.",
        "type": "internal"
    },
    "finishedAt": "2023-03-20T22:30:56.14873669Z",
    "indexUid": "addresses",
    "startedAt": "2023-03-14T05:42:21.639322047Z",
    "status": "failed",
    "type": "documentAdditionOrUpdate",
    "uid": 29436
}
```

Our workflow to update Meilisearch is as follows:
- Create 15 roughly equally sized JSON files with each row being a document to be indexed by Meilisearch.
  - Each file is ~1.7Gb in size with ~2.2million lines (documents) in each file
  - This results in a total number of documents of ~32,507,733
  - This is done to limit the amount of RAM needed for the job that sends the data to Meilisearch
- Use the Python SDK to load one file, and send to Meilisearch using `add_documents_in_batches`
  - We use a `batch_size` parameter of 10,000 (to keep the payload below 25Mb)
  - We let Meilisearch figure out how it wants to process/batch them as they come in. As our batch size is lower than the actual size of the batch we are sending, Meilisearch kicks in with its auto batching as we send it in.
- Once a file has been sent to Meilisearch we wait for `isIndexing` to be `false`
- Repeat the process for the next file, until all 32 million have been re-indexed by Meilisearch

This might seem a little bizzare but we make use of BigQuery, Dataflow and Cloud Composer to orchestrate our data into Meilisearch in GCP which is the reason we split the files up and do it in 15 "batches" over the weekend.

Every week we update the index by re-submitting this job (as so many documents can have little changes throughout the week it was simpler to just re-submit the entire job each week and let Meilisearch replace everything).

On upgrading to Meilisearch v1 I created the index from scratch running the process described above - this went fine.

On subsequent weeks when re-running the task we are seeing the error described above. The task stalls on the first file, and the tasks stay as `processing` and eventually (after around 6 days) they show this error.

CPU & RAM while this process is ongoing is static (see screenshot below). 

**To Reproduce**
Not sure how to reproduce since there's so much data involved. A typical document looks like (with keys & values redacted):

```json
{"id":"some text","b*e":"some text","p*e":null,"e*g":null,"t*e":null,"i*s":null,"n*s":null,"c*d":null,"i*w":null,"y*y":null,"e*e":null,"s*s":"some text","p**e":"some text","o*e":null,"d*e":null,"b*r":7,"b**e":null,"s*e":null,"d*y":"some text","d**y":null,"t**e":"some text","d**e":null,"p*n":"some text","_geo":{"lat":1.0,"lng":1.0}}
```

The size of each line as JSON is typically ~600-800bytes. The documents describe properties (so address, postcode etc along with internal fields describing the property (bedrooms, direction of garden etc) so unfortunately I cannot share them). The line which describes the postal address is almost always the longest (representing 10-15% of the total size of the document). The rest of the fields are roughly similar. Often a lot of the fields are null, but this varies greatly.

If needed I have the data in a GCS bucket, I can arrange private access to the data so the process can be replicated if this is required.

**Expected behavior**

Meilisearch on `v0.30.5` would process the documents as they come in and sucessfully update the index. This process took ~4-6 hours. I would expect `v1.0.2` to behave the same way.

**Screenshots**
CPU & RAM while the tasks are stuck in processing:
<img width="1626" alt="image" src="https://user-images.githubusercontent.com/53234158/227248705-8dd2aa5f-7c34-41e1-b9ce-8032243df9c6.png">

**Meilisearch version:** [e.g. v0.20.0]
`v1.0.2` Docker image using `getmeili/meilisearch`.

**Additional context**

Meilisearch is running on a GKE Autopilot cluster with 6vCPU and 20Gi RAM. There is a 150Gi PVC with a storage class `premium-rwo` (SSD) mapped to `/meili_data`.

We are not restrained by compute we can go much higher if needed but I found this to be the sweet spot on Meilisearch <v1.0.2.

We are using Meilisearch image `getmeili/meilisearch` with tag `v1.0.2`.

Meilisearch is deployed using a Helm chart we wrote internally which doesn't do anything unusual - it defines the deployment, the PVC, the service, a secret and an internal GCP load balancer. If needed I can share the chart.

I have looked through the issues and could not find a similar issue. Although it looks like point number 4 from [loiclec](https://github.com/loiclec) in this comment https://github.com/meilisearch/meilisearch/issues/3349#issuecomment-1411923608 may be related but this is a guess from my end.

**Comment by Kerollmops** [comment]
Hello @dtomlinson91,

Thank you very much for this well-described issue. We last had this LMDB error a while ago and couldn't reproduce it. We investigated a little bit and [discovered that it shouldn't happen with how we use this embedded key-value store](https://github.com/LMDB/lmdb/blob/3947014aed7ffe39a79991fa7fb5b234da47ad1a/libraries/liblmdb/mdb.c#L2270-L2301). We will open an issue on the OpenLDAP side to inform them about that. I hope they give an answer. We will see if they need a repro of the bug 🤞

Another strange behavior you describe in this issue: it froze when indexing. This is not normal behavior. I notice that the CPU is around 100% (maybe less than 1 CPU running), it makes me think that Meilisearch is trying to write to disk, but it's very slow. Would you mind measuring the disk activity (`iotop`) if it happens again? You talked about an SSD, but does it have maximum quotas? Like some GiB wrote by hour or something similar?

Thank you very much,
Have a nice day 🥣

**Comment by dtomlinson91** [comment]
Hey @Kerollmops thanks for replying 😁

If anyone needs the data/code used to replicate this I can arrange to share it to help with debugging.

Unfortunately Autopilot in GKE will not let you have privilged access to the nodes (since you have no access to the node at all it's all pay as you use) - so I cannot run `iotop`, I ran `kubestr` to benchmark the storage class which might be helpful instead.

The documentation for GKE/Compute Engine is a bit of a minefield, as far as I'm aware there are no GiB per hour quotas for `premium-rwo`, there is a higher tier where you pre-pay for iops but the details are abstracted away.

Result of `./kubestr fio -s premium-rwo`

```
PVC created kubestr-fio-pvc-2zvtg
W0323 16:43:59.312488    3074 warnings.go:70] Autopilot set default resource requests for Pod default/kubestr-fio-pod-*, as resource requests were not specified. See http://g.co/gke/autopilot-defaults
Pod created kubestr-fio-pod-m7f9q
Running FIO test (default-fio) on StorageClass (premium-rwo) with a PVC of Size (100Gi)
Elapsed time- 48.023934083s
FIO test results:

FIO version - fio-3.30
Global options - ioengine=libaio verify=0 direct=1 gtod_reduce=1

JobName: read_iops
  blocksize=4K filesize=2G iodepth=64 rw=randread
read:
  IOPS=3309.749756 BW(KiB/s)=13255
  iops: min=3088 max=3684 avg=3323.310303
  bw(KiB/s): min=12352 max=14736 avg=13293.517578

JobName: write_iops
  blocksize=4K filesize=2G iodepth=64 rw=randwrite
write:
  IOPS=2438.192627 BW(KiB/s)=9769
  iops: min=1790 max=3160 avg=2459.344727
  bw(KiB/s): min=7163 max=12640 avg=9837.758789

JobName: read_bw
  blocksize=128K filesize=2G iodepth=64 rw=randread
read:
  IOPS=2764.135986 BW(KiB/s)=354344
  iops: min=2493 max=2916 avg=2779.965576
  bw(KiB/s): min=319198 max=373248 avg=355848.125000

JobName: write_bw
  blocksize=128k filesize=2G iodepth=64 rw=randwrite
write:
  IOPS=2366.883545 BW(KiB/s)=303496
  iops: min=1976 max=2832 avg=2386.655273
  bw(KiB/s): min=252928 max=362538 avg=305513.406250

Disk stats (read/write):
  sdc: ios=103938/84636 merge=0/964 ticks=1176342/1065008 in_queue=2241351, util=68.315094%
  -  OK
```

**Comment by Kerollmops** [comment]
Thank you for your quick reply. I just [filed an issue](https://bugs.openldap.org/show_bug.cgi?id=10027) on the LMDB bug tracker.

**Comment by dtomlinson91** [comment]
@Kerollmops I know this isn't a priority due to the use case we have, but I can see that the issue you filed has stalled and I was wondering if there is any update?

I tried reducing the amount of data we send to Meilisearch at once. Instead of sending 1 NDJSON (1.7Gb) containing 2.2 million documents fifteen times, I sent 1 NDJSON (480Mb) with 660k documents 50 times (waiting between batch) and Meilisearch took 1-2 days to process a single batch (which is understandable from reading the docs, its very inefficient this way).

I've had to downgrade to v0.30.5 and it's rock solid. No issues with Meilisearch since rolling back.

I know this isn't a support forum, but is there anything you reccomend I try? Will increasing the RAM help? The DB size is ~120Gb but Meilisearch has only 20Gb available to it. Giving it 120Gb is impractical in GCP due to the cost of such a node, but would more RAM help?

Unfortunately LMDB goes over my head, despite looking at the issue I have no idea what I can do as a user of Meilisearch to help tune it.

**Comment by Kerollmops** [comment]
Hey @dtomlinson91,

It is indeed not a priority on our side. However, I would like to know if using [the latest version of the engine](https://github.com/meilisearch/meilisearch/releases/tag/v1.2.0) and [the new experimental `--experimental-reduce-indexing-memory-usage` flag](https://github.com/meilisearch/product/discussions/652) could help reduce the number of pages allocated in a single transaction?

In the next version of the engine, i.e., v1.3, [we removed an internal database](https://github.com/meilisearch/meilisearch/pull/3819). This will also reduce the number of pages in a single transaction which can, hopefully, fix your issue.

**Comment by dureuill** [comment]
Ping from triage:

- Severity: In some situations (changing settings), indexing will fail consistency on some workloads with no recourse. => Very high impact
  - Possible workaround, using [the new experimental `--experimental-reduce-indexing-memory-usage` flag](https://github.com/meilisearch/product/discussions/652) could prevent this issue. Needs more information on whether or not it works
- Size: if `--experimental-reduce-indexing-memory-usage` does not work, large to very large, requires structural changes to how we handle transactions

**Comment by Kerollmops** [comment]
Hey everybody 👋

We created [a prototype Docker Image that can be found on this PR](https://github.com/meilisearch/meilisearch/pull/4062#issue-1896612131). This prototype exposes a new experimental flag to limit the maximum number of auto-batched tasks that Meilisearch will process in one go.

In the future, we would like to better handle this `MDB_TXN_FULL` issue by automatically retrying to process half the batch, for example. It would be more accessible and transparent to users without making them aware of the engine's internals.

**Comment by dtomlinson91** [comment]
@Kerollmops Thanks for providing the experiment task limiting option. I tried it this week with various settings, fundamentally it seems its not the number of tasks that causes the error message we see, rather its the size of the batch itself.

We tried:

- Sending 15 batches of 100,000 to Meilisearch
  -  Not setting `MEILI_EXPERIMENTAL_LIMIT_BATCHED_TASKS` - Meilisearch processed the first task created, then processed the next 11 at once. This didn't work we get the same error.
  - Setting `MEILI_EXPERIMENTAL_LIMIT_BATCHED_TASKS=3` - Meilisearch processed the first task created, then processed the next 3, then the next etc. This did work, although it takes considerably longer to process the data this way (as is documented).
- Sending a single batch of 1.5million to Meilisearch (increasing the payload size)
  - Meilisearch creates and processes a single task - Didn't work, we see the same error

Because the error takes so long to manifest it's time consuming to debug, I believe giving Meilisearch fewer documents (say 500,000 instead of 1.5million) and (increasing the payload limit) submitting it as a single batch to Meilisearch would work, the only issue is 320/5 = 64 batches we'd have to send to Meilisearch, and it would need to finish in a weekend ready for Monday morning.

Since we are in the cloud and cost isn't really a factor, I think doubling the disk size and using the `/swap-indexes` endpoint makes more sense for us, since this error doesn't happen on a first index. I haven't tried this but it's a workaround to this if it does work, but ultimately processing a large number of document updates (whether its a single task or multiple tasks using batching) isn't working in v1 as it was in v0.
